File: jdbc/src/main/scala/moonbox/jdbc/MoonboxConnection.scala
/*-
 * <<
 * Moonbox
 * ==
 * Copyright (C) 2016 - 2018 EDP
 * ==
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * >>
 */

package moonbox.jdbc

import java.sql.{Blob, CallableStatement, Clob, Connection, DatabaseMetaData, NClob, PreparedStatement, SQLException, SQLWarning, SQLXML, Savepoint, Statement, Struct}
import java.util.{Locale, Properties}
import java.util.concurrent.Executor
import java.{sql, util}

import moonbox.client.JdbcClient
import moonbox.protocol.client._
import org.apache.commons.codec.digest.DigestUtils

class MoonboxConnection(url: String, props: Properties) extends java.sql.Connection {

  import moonbox.util.MoonboxJDBCUtils._

  private var jdbcSession: JdbcSession = _
  var statement: MoonboxStatement = _
  var closed: Boolean = _
  var database: String = _
  var networkTimeout: Int = 1000 * 60 * 3
  var DEFAULT_USER_CHECK_TIMEOUT = 1000 * 60 * 5
  var runMode = "local"

  private def isLocal: Boolean = {
    runMode.equalsIgnoreCase("local")
  }

  private def openSession(client: JdbcClient, username: String, password: String, newProps: Properties): Boolean = {
    val table = {
      val tb = newProps.getProperty("table")
      if (tb != null) tb
      else newProps.getProperty("collection")
    }
    if (database == null) {
      database = newProps.getProperty(DB_NAME)
    }
    val message = OpenSessionInbound(null, Some(database), isLocal).setId(client.genMessageId)
    val resp = client.sendAndReceive(message)
    resp match {
      case OpenSessionOutbound(_, None) =>
        initSession(client, database, table, username, password, isLocal, newProps)
        true
      case other =>
        client.close()
        throw new SQLException(s"Get moonbox connection failed: $other")
    }
  }

  private def clientGen(host: String, port: Int, username: String, password: String): JdbcClient = {
    val client = new JdbcClient(host, port)
    client.connect()
    val (newClient, errorMessage) = if (client.isConnected) {
      val loginOutbound = client.sendAndReceive(LoginInbound(username, password).setId(client.genMessageId))
      loginOutbound match {
        case LoginOutbound(Some(_), None) =>
          val resp = client.sendAndReceive(RequestAccessInbound(None, isLocal).setId(client.genMessageId), DEFAULT_USER_CHECK_TIMEOUT)
          client.close()
          resp match {
            case RequestAccessOutbound(Some(address), None) =>
              val (host, port) = splitHostPort(address)
              (Some(new JdbcClient(host, port)), None)
            case other => (None, Some(s"Get moonbox connection failed: $other"))
          }
        case other: LoginOutbound => (None, Some(s"Get moonbox connection failed: error=${other.error}"))
        case e => (None, Some(s"Get moonbox connection failed: $e"))
      }
    } else (None, Some("Cannot connect to moonbox."))
    newClient.getOrElse(throw new SQLException(errorMessage.orNull))
  }

  def userCheck(): Boolean = {
    var flag = false
    val newProps = parseURL(url, props)
    val username = newProps.getProperty(USER_KEY)
    val pwd = newProps.getProperty(PASSWORD_KEY)
    runMode = Option(newProps.getProperty(MODE_KEY)).getOrElse("local")
    val (host, port) = parseHostsAndPorts(newProps.getProperty(HOSTS_AND_PORTS)).map { case (h, p) => (h, p.toInt) }.head
    val client = clientGen(host, port, username, pwd)
    client.connect()
    if (client.isConnected) {
      val resp = client.sendAndReceive(LoginInbound(username, pwd).setId(client.genMessageId), DEFAULT_USER_CHECK_TIMEOUT)
      resp match {
        case LoginOutbound(_, None) =>
          flag = openSession(client, username, pwd, newProps)
        case other =>
          client.close()
          throw new SQLException(s"Get moonbox connection failed: $other")
      }
    }
    flag
  }

  private def initSession(jdbcClient: JdbcClient, database: String, table: String, username: String, pwd: String, isLocal: Boolean, props: Properties): Unit = {
    if (pwd != null && pwd.length > 0)
      jdbcSession = JdbcSession(jdbcClient, database, table, username, DigestUtils.md5Hex(pwd), props, isLocal)
    else
      jdbcSession = JdbcSession(jdbcClient, database, table, username, pwd, props, isLocal)
    closed = false
  }

  def getSession(): JdbcSession = jdbcSession

  override def commit(): Unit = throw new SQLException("Unsupported")

  override def getHoldability: Int = 1

  override def setCatalog(catalog: String): Unit = {
    this.createStatement().executeQuery(s"use $catalog")
    database = catalog
    jdbcSession = jdbcSession.copy(database = catalog)
  }

  override def setHoldability(holdability: Int): Unit = {}

  override def prepareStatement(sql: String): PreparedStatement = throw new SQLException("Unsupported")

  override def prepareStatement(sql: String, resultSetType: Int, resultSetConcurrency: Int): PreparedStatement = throw new SQLException("Unsupported")

  override def prepareStatement(sql: String, resultSetType: Int, resultSetConcurrency: Int, resultSetHoldability: Int): PreparedStatement = throw new SQLException("Unsupported")

  override def prepareStatement(sql: String, autoGeneratedKeys: Int): PreparedStatement = throw new SQLException("Unsupported")

  override def prepareStatement(sql: String, columnIndexes: Array[Int]): PreparedStatement = throw new SQLException("Unsupported")

  override def prepareStatement(sql: String, columnNames: Array[String]): PreparedStatement = throw new SQLException("Unsupported")

  override def createClob(): Clob = throw new SQLException("Unsupported")

  override def setSchema(schema: String): Unit = {}

  override def setClientInfo(name: String, value: String): Unit = {}

  override def setClientInfo(properties: Properties): Unit = {}

  override def createSQLXML(): SQLXML = throw new SQLException("Unsupported")

  override def getCatalog: String = database

  override def createBlob(): Blob = throw new SQLException("Unsupported")

  def checkClosed(): Unit = {
    if (isClosed()) {
      throw new SQLException("Connection is already closed.")
    }
  }

  override def createStatement(): Statement = {
    checkClosed()
    statement = new MoonboxStatement(this)
    statement
  }

  override def createStatement(resultSetType: Int, resultSetConcurrency: Int): Statement = createStatement()

  override def createStatement(resultSetType: Int, resultSetConcurrency: Int, resultSetHoldability: Int): Statement = createStatement()

  override def abort(executor: Executor): Unit = this.close()

  override def setAutoCommit(autoCommit: Boolean): Unit = {}

  override def getMetaData: DatabaseMetaData = {
    new MoonboxDatabaseMetaData(this)
  }

  override def setReadOnly(readOnly: Boolean): Unit = {}

  override def prepareCall(sql: String): CallableStatement = throw new SQLException("Unsupported")

  override def prepareCall(sql: String, resultSetType: Int, resultSetConcurrency: Int): CallableStatement = throw new SQLException("Unsupported")

  override def prepareCall(sql: String, resultSetType: Int, resultSetConcurrency: Int, resultSetHoldability: Int): CallableStatement = throw new SQLException("Unsupported")

  override def setTransactionIsolation(level: Int): Unit = {}

  override def getWarnings: SQLWarning = {
    throw new SQLException("Unsupported")
  }

  override def releaseSavepoint(savepoint: Savepoint): Unit = throw new SQLException("Unsupported")

  override def nativeSQL(sql: String): String = throw new SQLException("Unsupported")

  override def isReadOnly: Boolean = false

  override def createArrayOf(typeName: String, elements: Array[AnyRef]): sql.Array = throw new SQLException("Unsupported")

  override def setSavepoint(): Savepoint = {
    throw new SQLException("Unsupported")
  }

  override def setSavepoint(name: String): Savepoint = {
    throw new SQLException("Unsupported")
  }

  override def close(): Unit = {
    if (statement != null && !statement.isClosed) {
      statement.close()
    }
    statement = null
    if (jdbcSession != null || !jdbcSession.isClosed()) {
      jdbcSession.close()
    }
    jdbcSession = null
    closed = true
  }

  override def createNClob(): NClob = throw new SQLException("Unsupported")

  override def rollback(): Unit = {}

  override def rollback(savepoint: Savepoint): Unit = {}

  override def setNetworkTimeout(executor: Executor, milliseconds: Int): Unit = {
    networkTimeout = milliseconds
  }

  override def setTypeMap(map: util.Map[String, Class[_]]): Unit = {
  }

  override def isValid(timeout: Int): Boolean = {
    if (jdbcSession == null || isClosed()) false else true
  }

  override def getAutoCommit: Boolean = false

  override def clearWarnings(): Unit = {}

  override def getSchema: String = throw new SQLException("Unsupported")

  override def getNetworkTimeout: Int = {
    networkTimeout
  }

  override def isClosed(): Boolean = {
    synchronized {
      if (jdbcSession == null || jdbcSession.isClosed()) {
        closed = true
      }
      closed
    }
  }

  override def getTransactionIsolation: Int = {
    Connection.TRANSACTION_NONE
  }

  override def createStruct(typeName: String, attributes: Array[AnyRef]): Struct = throw new SQLException("Unsupported")

  override def getClientInfo(name: String): String = throw new SQLException("Unsupported")

  override def getClientInfo: Properties = throw new SQLException("Unsupported")

  override def getTypeMap: util.Map[String, Class[_]] = throw new SQLException("Unsupported")

  override def unwrap[T](iface: Class[T]): T = {
    if (isWrapperFor(iface))
      this.asInstanceOf[T]
    else
      throw new SQLException("unwrap exception")
  }

  override def isWrapperFor(iface: Class[_]): Boolean = iface != null && iface.isAssignableFrom(getClass)
}

File: jdbc/src/main/java/moonbox/jdbc/MoonboxConnection.java
package moonbox.jdbc;

import moonbox.jdbc.util.Utils;

import java.sql.*;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.Executor;

public class MoonboxConnection implements Connection {
  private static final String DEFAULT_MASTER_HOST = "localhost";
  private static final int DEFAULT_MASTER_PORT = 10010;
  private static final int DEFAULT_CONNECT_TIMEOUT = 20 * 1000;
  private static final String DEFAULT_DATABASE = "default";
  private Properties info;
  private MoonboxClient client;
  private String masterHost;
  private int masterPort;
  private int connectTimeout;
  private String database;

  private DatabaseMetaData dbmd;

  public MoonboxConnection(String url, Properties info) throws SQLException {
    this.info = Utils.parseURL(url, info);
    this.database = this.info.getProperty("database", DEFAULT_DATABASE);
    this.masterHost = this.info.getProperty("host", DEFAULT_MASTER_HOST);
    this.masterPort = Integer.valueOf(this.info.getProperty("port", String.valueOf(DEFAULT_MASTER_PORT)));
    this.connectTimeout = Integer.valueOf(this.info.getProperty("connectTimeout", String.valueOf(DEFAULT_CONNECT_TIMEOUT)));
    this.client = createClient();
    this.dbmd = getMetaData(false);
  }

  private MoonboxClient createClient() throws SQLException {
    String user = this.info.getProperty("user");
    String password = this.info.getProperty("password");
    String appType = this.info.getProperty("apptype");
    String appName = this.info.getProperty("appname", null);
    Map<String, String> config = new HashMap<>();
    config.put("database", database);
    try {
      return new MoonboxClient(masterHost,
          masterPort, connectTimeout, user, password, appType, appName, config);
    } catch (Exception e) {
      throw new SQLException(e);
    }
  }

  MoonboxClient getClient() {
    return this.client;
  }

  int getConnectTimeout() {
    return this.connectTimeout;
  }

  @Override
  public Statement createStatement() throws SQLException {
    checkClosed();
    return new MoonboxStatement(this, this.database);
  }

  @Override
  public PreparedStatement prepareStatement(String sql) throws SQLException {
    checkClosed();
    return new MoonboxPrepareStatement();
  }

  @Override
  public CallableStatement prepareCall(String sql) throws SQLException {
    return null;
  }

  @Override
  public String nativeSQL(String sql) throws SQLException {
    return null;
  }

  @Override
  public void setAutoCommit(boolean autoCommit) throws SQLException {

  }

  @Override
  public boolean getAutoCommit() throws SQLException {
    return false;
  }

  @Override
  public void commit() throws SQLException {

  }

  @Override
  public void rollback() throws SQLException {

  }

  @Override
  public void close() throws SQLException {
    try {
      client.close();
    } catch (Exception e) {
      throw new SQLException(e);
    }
  }

  @Override
  public boolean isClosed() throws SQLException {
    return client.isClose();
  }

  @Override
  public DatabaseMetaData getMetaData() throws SQLException {
    return getMetaData(true);
  }

  private DatabaseMetaData getMetaData(boolean checkClosed) throws SQLException {
    if (checkClosed) {
      checkClosed();
    }
    return new MoonboxDatabaseMetaData(this, this.database);
  }

  public void checkClosed() throws SQLException {
    if (this.client.isClose()) {
      throw new SQLException("No operations allowed after connection closed.");
    }
  }

  @Override
  public void setReadOnly(boolean readOnly) throws SQLException {

  }

  @Override
  public boolean isReadOnly() throws SQLException {
    return true;
  }

  @Override
  public void setCatalog(String catalog) throws SQLException {
    createStatement().executeQuery("USE " + catalog);
    this.database = catalog;
  }

  @Override
  public String getCatalog() throws SQLException {
    return this.database;
  }

  @Override
  public void setTransactionIsolation(int level) throws SQLException {

  }

  @Override
  public int getTransactionIsolation() throws SQLException {
    return 0;
  }

  @Override
  public SQLWarning getWarnings() throws SQLException {
    return null;
  }

  @Override
  public void clearWarnings() throws SQLException {

  }

  @Override
  public Statement createStatement(int resultSetType, int resultSetConcurrency) throws SQLException {
    return createStatement();
  }

  @Override
  public PreparedStatement prepareStatement(String sql, int resultSetType, int resultSetConcurrency) throws SQLException {
    return prepareStatement(sql);
  }

  @Override
  public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency) throws SQLException {
    return null;
  }

  @Override
  public Map<String, Class<?>> getTypeMap() throws SQLException {
    return null;
  }

  @Override
  public void setTypeMap(Map<String, Class<?>> map) throws SQLException {

  }

  @Override
  public void setHoldability(int holdability) throws SQLException {

  }

  @Override
  public int getHoldability() throws SQLException {
    return 0;
  }

  @Override
  public Savepoint setSavepoint() throws SQLException {
    return null;
  }

  @Override
  public Savepoint setSavepoint(String name) throws SQLException {
    return null;
  }

  @Override
  public void rollback(Savepoint savepoint) throws SQLException {

  }

  @Override
  public void releaseSavepoint(Savepoint savepoint) throws SQLException {

  }

  @Override
  public Statement createStatement(int resultSetType, int resultSetConcurrency, int resultSetHoldability) throws SQLException {
    return createStatement();
  }

  @Override
  public PreparedStatement prepareStatement(String sql, int resultSetType, int resultSetConcurrency, int resultSetHoldability) throws SQLException {
    return prepareStatement(sql);
  }

  @Override
  public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency, int resultSetHoldability) throws SQLException {
    return null;
  }

  @Override
  public PreparedStatement prepareStatement(String sql, int autoGeneratedKeys) throws SQLException {
    return prepareStatement(sql);
  }

  @Override
  public PreparedStatement prepareStatement(String sql, int[] columnIndexes) throws SQLException {
    return prepareStatement(sql);
  }

  @Override
  public PreparedStatement prepareStatement(String sql, String[] columnNames) throws SQLException {
    return prepareStatement(sql);
  }

  @Override
  public Clob createClob() throws SQLException {
    return null;
  }

  @Override
  public Blob createBlob() throws SQLException {
    return null;
  }

  @Override
  public NClob createNClob() throws SQLException {
    return null;
  }

  @Override
  public SQLXML createSQLXML() throws SQLException {
    return null;
  }

  @Override
  public boolean isValid(int timeout) throws SQLException {
    return false;
  }

  @Override
  public void setClientInfo(String name, String value) throws SQLClientInfoException {

  }

  @Override
  public void setClientInfo(Properties properties) throws SQLClientInfoException {

  }

  @Override
  public String getClientInfo(String name) throws SQLException {
    return this.info.getProperty(name);
  }

  @Override
  public Properties getClientInfo() throws SQLException {
    return this.info;
  }

  @Override
  public Array createArrayOf(String typeName, Object[] elements) throws SQLException {
    return null;
  }

  @Override
  public Struct createStruct(String typeName, Object[] attributes) throws SQLException {
    return null;
  }

  @Override
  public void setSchema(String schema) throws SQLException {
    setCatalog(schema);
  }

  @Override
  public String getSchema() throws SQLException {
    return getCatalog();
  }

  @Override
  public void abort(Executor executor) throws SQLException {

  }

  @Override
  public void setNetworkTimeout(Executor executor, int milliseconds) throws SQLException {

  }

  @Override
  public int getNetworkTimeout() throws SQLException {
    return 0;
  }

  @Override
  public <T> T unwrap(Class<T> iface) throws SQLException {
    return null;
  }

  @Override
  public boolean isWrapperFor(Class<?> iface) throws SQLException {
    return false;
  }
}
